packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -3123,28 +3123,40 @@ namespace ColMapFunctors
     bitset_t present;
   };
 
-  template<typename GO, typename GView, typename const_bitset_t>
+  template<typename LO, typename GO, typename GView, typename const_bitset_t, typename LocalMapType>
   struct ListGIDs
   {
-    ListGIDs(GO minGID_, GView& gidList_, const_bitset_t& present_)
-      : minGID(minGID_), gidList(gidList_), present(present_)
+    ListGIDs(GO minGID_, GView& gidList_, const_bitset_t& present_, const LocalMapType& localDomainMap_)
+      : minGID(minGID_), doingRemotes(false), gidList(gidList_), present(present_), localDomainMap(localDomainMap_)
     {}
 
     KOKKOS_INLINE_FUNCTION void operator()(const GO i, GO& lcount, const bool finalPass) const
     {
-      if(finalPass && present.test(i))
-      {
-        gidList(lcount) = minGID + i;
-      }
-      if(present.test(i))
+      bool isRemote = localDomainMap.getLocalElement(i + minGID) == ::Tpetra::Details::OrdinalTraits<LO>::invalid();
+      if(present.test(i) && doingRemotes == isRemote)
       {
+        if(finalPass)
+        {
+          if(doingRemotes)
+          {
+            //remotes are inserted at the end
```

Comment:
You might actually want to use the existing Tpetra function `makeColMap` to make the column Map, rather than trying to reimplement it (but not correctly) here.  On the other hand, note that `makeColMap` needs to do MPI communication by default.  It would be great if you already had that data before entering your function.

Reply:
@mhoemmen What is the issue you see with this col map construction? I would be fine with using makeColMap, but if I go through every row in serial on host and do insertGlobalEntries() with the global colinds and values I have, the speed benefit from using the KokkosKernels spadd is mostly negated. Is there a way to construct a globally indexed, packed 1D storage CrsGraph? Then I would be done. I wouldn't even have to make a column map then if fill complete wasn't requested (which it isn't in Xpetra::TwoMatrixAdd, which seems to be the most common user of this add()).

Reply:
@brian-kelley wrote:
> Do the remote GIDs need to be in a certain order?

When you let the CrsGraph or CrsMatrix construct its own column Map at fillComplete, then the remote GIDs on each process are sorted first by owning process rank in the domain Map.  The GID order affects the LIDs.  Some tests might just check the LIDs and/or column Map for equality directly.  I'm not saying your algorithm is incorrect, just that you might be producing a different column Map in a way that makes the tests grumpy.

> Is there a way to construct a globally indexed, packed 1D storage CrsGraph?

Do you mean "give the CrsGraph's constructor all the global indices at once, as packed 1-D storage"?  Not currently; CrsGraph doesn't have that constructor.

Reply:
@mhoemmen Thank you! That function would solve half the problem. The other half is the case where we don't have a domain/range map, and ``add()`` produces a non-fill complete C. Could I make a version ``makeCrsGraphFromGlobalIndices`` that just takes a ``StaticCrsGraph<GO>`` and row map? Could these just be static functions inside CrsGraph?

Reply:
@brian-kelley wrote:
> Could these just be static functions inside CrsGraph?

[C++ Core Guidelines C.4](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-member): "Make a function a member only if it needs direct access to the representation of a class."

> add is going to fillComplete and then resumeFill to let the user sum into the existing values - I think that's the best I can do.

What's the intent of "add and return a non-fill-complete matrix"?  Would users be surprised that the resulting matrix is locally indexed and has a column Map?